### PR TITLE
feature(generator): check for changes in golden files to indicate generator plugin behavior changes

### LINKED ISSUE #5832: generator: ensure existing libraries are regenerated when microgenerator changes
Currently, we can detect when the protos are updated and can issue a warning to run the microgenerator, but we lack a mechanism to detect/execute generation of libraries when the behavior of the microgenerator changes but the protos remain the same.

### PR DESCRIPTION
**Pull request by scotthart**
fixes #5832 

Compare md5 hashes of golden files across commits to determine if changes have been made to the emitted output of the generator that would require existing libraries to be regenerated. By checking the observable differences in the generator output, we can limit library regeneration to only when necessary.

Also adds missing mock_golden_kitchen_sink file to build files.

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5861)

### CODE REVIEW

File: generator/generate-libraries.sh [hunk 1]
```diff
@@ -83,4 +83,7 @@ GOOGLEAPIS_HASH_LAST_USED=$BAZEL_DEPS_GOOGLEAPIS_HASH
 readonly GOOGLEAPIS_HASH_LAST_USED
 EOF
 
+io::log_yellow "Update ci/etc/generator-golden-md5-hashes.sh."
+find generator/integration_tests/golden -name '*.gcpcxx.*' -print0 | sort -z | xargs -0 md5sum >>"${PROJECT_ROOT}"/ci/etc/generator-golden-md5-hashes.md5
```
Comment (devbww) [inline generator/generate-libraries.sh:87]: Break long line perhaps?

File: ci/check-golden-md5-hashes.sh [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+source "$(dirname "$0")/lib/init.sh"
+source module lib/io.sh
+
+if [[ "${CHECK_GENERATED_CODE_HASH}" != "yes" ]]; then
+  echo "Skipping the generated code hash check as it is disabled for this build."
+  exit 0
+fi
+
+cd "${PROJECT_ROOT}"
```
Comment (devbww) [inline ci/check-golden-md5-hashes.sh:26]: Do we need the `${PROJECT_ROOT}/` prefixes below if we're going to `cd` (or vice versa)?

File: ci/kokoro/docker/build-in-docker-bazel.sh [hunk 3]
```diff
@@ -106,6 +106,13 @@ io::log_yellow "Verify googleapis commit hash for generated code"
   ./ci/check-generated-code-hash.sh
 )
 
+echo "================================================================"
+io::log_yellow "Verify generator golden file md5 hashes"
+(
+  cd "${PROJECT_ROOT}"
```
Comment (devbww) [inline ci/kokoro/docker/build-in-docker-bazel.sh:112]: Do we need this if `check-golden-md5-hashes.sh` is going to heed `${PROJECT_ROOT}` itself (or vice versa)?

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5861)* status: 0 of 6 files reviewed, 3 unresolved discussions (waiting on @devbww)

___
*[ci/check-golden-md5-hashes.sh, line 26 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5861#-MThXUXN4gk20QplhiO8-r2-26:-MTv_H1pAXVRQhEoELmy:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/5a2343533cc8dc989182017c2dc88170dfe61a2f/ci/check-golden-md5-hashes.sh#L26)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

Do we need the `${PROJECT_ROOT}/` prefixes below if we're going to `cd` (or vice versa)?
</blockquote></details>

Done.
___
*[ci/kokoro/docker/build-in-docker-bazel.sh, line 112 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5861#-MThXUXN4gk20QplhiO4-r2-112:-MTvcV5910hqwybF2RpY:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/5a2343533cc8dc989182017c2dc88170dfe61a2f/ci/kokoro/docker/build-in-docker-bazel.sh#L112)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

Do we need this if `check-golden-md5-hashes.sh` is going to heed `${PROJECT_ROOT}` itself (or vice versa)?
</blockquote></details>

Done.
___
*[generator/generate-libraries.sh, line 87 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5861#-MThXUXN4gk20QplhiO5-r2-87:-MTv_JwKAKmAqc1xeYvh:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/5a2343533cc8dc989182017c2dc88170dfe61a2f/generator/generate-libraries.sh#L87)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

Break long line perhaps?
</blockquote></details>

Done.

File: ci/kokoro/docker/build-in-docker-bazel.sh [hunk 4]
```diff
@@ -102,15 +102,13 @@ fi
 echo "================================================================"
 io::log_yellow "Verify googleapis commit hash for generated code"
 (
```
Comment (devbww) [inline ci/kokoro/docker/build-in-docker-bazel.sh:104]: Now we don't need the sub-shell.

File: ci/kokoro/docker/build-in-docker-bazel.sh [hunk 5]
```diff
@@ -102,15 +102,13 @@ fi
...
-  cd "${PROJECT_ROOT}"
-  ./ci/check-generated-code-hash.sh
+  "${PROJECT_ROOT}"/ci/check-generated-code-hash.sh
 )
 
 echo "================================================================"
 io::log_yellow "Verify generator golden file md5 hashes"
 (
```
Comment (devbww) [inline ci/kokoro/docker/build-in-docker-bazel.sh:110]: Now we don't need the sub-shell.

File: generator/generate-libraries.sh [hunk 6]
```diff
@@ -60,7 +60,7 @@ for proto in "${product_path_proto_path_to_generate[@]}"; do
   find "${product_path}" \( -name '*.cc' -o -name '*.h' \) -exec clang-format -i {} \;
 done
 
-io::log_yellow "Update generator-googlapis-commit-hash.sh file."
+io::log_yellow "Update generator-googlapis-commit-hash. file."
```
Comment (devbww) [inline generator/generate-libraries.sh:63]: Did you mean to lose the `sh`?

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5861)* status: 0 of 6 files reviewed, 6 unresolved discussions (waiting on @devbww)

___
*[ci/kokoro/docker/build-in-docker-bazel.sh, line 104 at r3](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5861#-MThXUXN4gk20QplhiO4-r3-104:-MTvmgAE85v_g6L9SDDW:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/b2a6aff9a523ecde289a39022500fafe7a97cc8c/ci/kokoro/docker/build-in-docker-bazel.sh#L104)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

Now we don't need the sub-shell.
</blockquote></details>

Done.
___
*[ci/kokoro/docker/build-in-docker-bazel.sh, line 110 at r3](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5861#-MThXUXN4gk20QplhiO4-r3-110:-MTvmgylBAZfYbb7RFvL:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/b2a6aff9a523ecde289a39022500fafe7a97cc8c/ci/kokoro/docker/build-in-docker-bazel.sh#L110)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

Now we don't need the sub-shell.
</blockquote></details>

Done.
___
*[generator/generate-libraries.sh, line 63 at r3](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5861#-MThXUXN4gk20QplhiO5-r3-63:-MTvmiZh22daQdyb7ORG:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/b2a6aff9a523ecde289a39022500fafe7a97cc8c/generator/generate-libraries.sh#L63)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

Did you mean to lose the `sh`?
</blockquote></details>

Done.

**Review by devbww — APPROVED** [review]

**Pull request merged**